### PR TITLE
extend build stage from CI/CD container build

### PR DESCRIPTION
**Pull request by rasolca**
Beside simplifying the maintenance of the CI yaml file,
this approach automatically retries when pushing to jfrog fails.

The intel apt-get repository gpg key has been updated as well.

### CODE REVIEW

**Comment by rasolca** [comment]
cscs-ci run

**Comment by rasolca** [comment]
cscs-ci run

**Comment by rasolca** [comment]
cscs-ci run

**Comment by rasolca** [comment]
cscs-ci run

**Comment by rasolca** [comment]
cscs-ci run

**Comment by rasolca** [comment]
cscs-ci run

**Comment by rasolca** [comment]
cscs-ci run

**Comment by rasolca** [comment]
cscs-ci run

**Comment by rasolca** [comment]
cscs-ci run

File: ci/cpu/clang12_release.yml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  needs:
+    - cpu clang12 release deps
+  variables:
+    DOCKERFILE: ci/docker/deploy.Dockerfile
+    DEPLOY_BASE_IMAGE: docker.io/ubuntu:22.04
+    EXTRA_APTGET_DEPLOY: "glibc-tools jq strace"
+    # glibc-tools is needed for libSegFault on ubuntu:22.04
+    DEPLOY_IMAGE: $CSCS_REGISTRY_PATH/release-cpu-clang12/deploy:$CI_COMMIT_SHA
+    SLURM_CONSTRAINT: mc
+    THREADS_PER_NODE: 72
```
Comment (msimberg) [inline ci/cpu/clang12_release.yml:31]: I guess we could move a few more variables up to `build_*common`:
- `DOCKERFILE`
- `BASE_IMAGE`
- `COMMON_SPACK_ENVIRONMENT`
assuming it's possible to override variables when needed (which it seems like it is). Things like `CXXSTD` could at least default to `17` and `USE_ROCBLAS` to `OFF`.

Would it also make sense to define a few reusable blocks like `mc_constraint` and `gpu_constraint`? As far as I can tell it's possible to use multiple base configurations in `.extends` (https://docs.gitlab.com/ee/ci/yaml/yaml_optimization.html#merge-details). Obviously a "reusable base configuration" with only one variable doesn't make much sense, so this is only useful if we can group more than one variable.

I admit I haven't tried this so I'm not 100% sure it works.

**Comment by rasolca** [comment]
cscs-ci run

File: ci/cpu/clang12_release.yml (hunk 1 above)
Comment (rasolca) [reply ci/cpu/clang12_release.yml]: Good idea. Let me try.

**Comment by rasolca** [comment]
Note: To avoid mixing multiple coverage reports, I will keep only one coverage build (cuda gcc11).
      The other coverage builds will converted to debug builds.

**Comment by gulivarese** [comment]
cscs-ci run

**Comment by gulivarese** [comment]
cscs-ci run

**Comment by gulivarese** [comment]
cscs-ci run

**Comment by gulivarese** [comment]
cscs-ci run

**Comment by gulivarese** [comment]
cscs-ci run

**Comment by gulivarese** [comment]
cscs-ci run

**Comment by gulivarese** [comment]
cscs-ci run

**Comment by gulivarese** [comment]
cscs-ci run

**Comment by gulivarese** [comment]
cscs-ci run

**Comment by gulivarese** [comment]
cscs-ci run

**Comment by gulivarese** [comment]
cscs-ci run

**Comment by gulivarese** [comment]
cscs-ci run

**Comment by gulivarese** [comment]
cscs-ci run

**Comment by gulivarese** [comment]
cscs-ci run

**Comment by gulivarese** [comment]
cscs-ci run

**Comment by gulivarese** [comment]
cscs-ci run

**Comment by gulivarese** [comment]
cscs-ci run

**Comment by gulivarese** [comment]
cscs-ci run

**Comment by gulivarese** [comment]
cscs-ci run

**Comment by gulivarese** [comment]
cscs-ci run

File: ci/cpu/clang13_release_cxx20.yml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    DEPLOY_IMAGE: $CSCS_REGISTRY_PATH/release-cpu-clang13-20/deploy:$CI_COMMIT_SHA
+
+cpu clang13 release cxx20 test:
+  extends: .run_common
+  needs:
+    - cpu clang13 release cxx20 build
+  trigger:
+    include:
+      - artifact: pipeline.yml
+        job: cpu clang13 release cxx20 build
```
Comment (rasolca) [inline ci/cpu/clang13_release_cxx20.yml:30]: The new line character is missing at the end of the file (see https://github.com/eth-cscs/DLA-Future/blob/master/misc/coding.md#vertical-spacing).
This is the problem reported by `check-format`

See https://github.com/eth-cscs/DLA-Future/actions/runs/6393725766/job/17353554378?pr=988#step:8:22

Other files has the same problem.

File: ci/.gitlab-ci.yml [hunk 3]
```diff
@@ -1,442 +1,101 @@
...
+#        job: cuda codecov build gcc11
+#
+#cuda codecov test gcc11 scalapack:
+#  extends: .run_common
+#  needs:
+#    - cuda codecov build gcc11 scalapack
+#  trigger:
+#    include:
+#      - artifact: pipeline.yml
+#        job: cuda codecov build gcc11 scalapack
```
Comment (rasolca) [inline ci/.gitlab-ci.yml:101]: Here you have too many empty lines at the end of the file.

File: ci/cpu/clang12_release.yml [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    EXTRA_APTGET: "clang-12 libomp-12-dev"
+    COMPILER: clang@12.0.1
+    USE_MKL: "ON"
+    SPACK_ENVIRONMENT: ci/docker/release-cpu.yaml
+    BUILD_IMAGE: $CSCS_REGISTRY_PATH/release-cpu-clang12/build
+
+cpu clang12 release build:
+  extends:
+    - .build_common
+    - .build_for_daint-mc
```
Comment (rasolca) [inline ci/cpu/clang12_release.yml:16]: Cuda tests should be run on daint-gpu.
```suggestion
    - .build_for_daint-gpu
```
(Applies to other cuda tests)

File: ci/rocm/clang14+rocm-5.3.3_release.yml [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    EXTRA_APTGET: "clang-14 libomp-14-dev rocblas rocblas-dev rocsolver rocsolver-dev rocprim-dev rocthrust-dev llvm-amdgpu rocm-device-libs"
+    COMPILER: clang@14.0.0
+    USE_ROCBLAS: "ON"
+    SPACK_ENVIRONMENT: ci/docker/release-rocm533.yaml
+    BUILD_IMAGE: $CSCS_REGISTRY_PATH/release-rocm-clang14/build
+
+rocm clang14+rocm-5.3.3 release build:
+  extends:
+    - .build_common
+    - .build_for_daint-mc
```
Comment (rasolca) [inline ci/rocm/clang14+rocm-5.3.3_release.yml:17]: Tests of this config are not run yet, but they will for sure not run on daint.
In this case is maybe better to disable the creation of the pipeline with:
```suggestion
  after_script: null
```

**Comment by gulivarese** [comment]
cscs-ci run

**Comment by gulivarese** [comment]
cscs-ci run

**Comment by gulivarese** [comment]
cscs-ci run

File: ci/.gitlab-ci.yml (hunk 3 above)
Comment (gulivarese) [reply ci/.gitlab-ci.yml]: Fixed

File: ci/cpu/clang12_release.yml (hunk 4 above)
Comment (gulivarese) [reply ci/cpu/clang12_release.yml:16]: Fixed

**Comment by gulivarese** [comment]
cscs-ci run

**Comment by gulivarese** [comment]
cscs-ci run

**Comment by gulivarese** [comment]
cscs-ci run

**Comment by gulivarese** [comment]
cscs-ci run

**Comment by gulivarese** [comment]
cscs-ci run

**Comment by gulivarese** [comment]
cscs-ci run

**Comment by gulivarese** [comment]
cscs-ci run

**Comment by gulivarese** [comment]
cscs-ci run

**Comment by gulivarese** [comment]
cscs-ci run

**Comment by gulivarese** [comment]
cscs-ci run

File: ci/cuda/gcc11_codecov_scalapack.yml [hunk 6]
```diff
@@ -0,0 +1,29 @@
+include:
+  - local: 'ci/common-ci.yml'
+
+cuda gcc11 codecov scalapack deps:
+  extends: .build_deps_common
+  variables:
+    BASE_IMAGE: docker.io/nvidia/cuda:11.7.1-devel-ubuntu22.04
+    COMPILER: gcc@11.2.0
+    SPACK_ENVIRONMENT: $CSCS_REGISTRY_PATH/cuda-gcc11-debug/build
```
Comment (rasolca) [inline ci/cuda/gcc11_codecov_scalapack.yml:9]: This value is surely wrong.

File: ci/cuda/gcc11_codecov_scalapack.yml [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    BUILD_IMAGE: $CSCS_REGISTRY_PATH/codecov-cuda-gcc11/build
```
Comment (rasolca) [inline ci/cuda/gcc11_codecov_scalapack.yml:10]:
```suggestion
    BUILD_IMAGE: $CSCS_REGISTRY_PATH/cuda-gcc11-debug/build
```

**Comment by gulivarese** [comment]
cscs-ci run

**Comment by gulivarese** [comment]
cscs-ci run

**Comment by gulivarese** [comment]
cscs-ci run

**Comment by gulivarese** [comment]
cscs-ci run

**Comment by gulivarese** [comment]
cscs-ci run

**Comment by gulivarese** [comment]
cscs-ci run

**Comment by rasolca** [comment]
cscs-ci run

File: ci/common-ci.yml [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+        "USE_MKL",
+        "USE_ROCBLAS",
+        "NUM_PROCS=$NUM_CORES_BUILD_DLAF"
+      ]'
+    # default configuration variables
+    # can be overwritten in the configuration as needed
+    DOCKERFILE: ci/docker/deploy.Dockerfile
+    DEPLOY_BASE_IMAGE: docker.io/ubuntu:22.04
+    EXTRA_APTGET_DEPLOY: "glibc-tools jq strace"
+    # glibc-tools is needed for libSegFault on ubuntu:22.04
```
Comment (msimberg) [inline ci/common-ci.yml:86]: Does the comment make more sense above `EXTRA_APTGET_DEPLOY`?
```suggestion
    # glibc-tools is needed for libSegFault on ubuntu:22.04
    EXTRA_APTGET_DEPLOY: "glibc-tools jq strace"
```
Then again, if we need a comment for `glibc-tools`, do we need one for `jq` and `strace`?

File: ci/rocm/clang14+rocm-5.3.3_release.yml [hunk 9]
```diff
@@ -0,0 +1,23 @@
+include:
+  - local: 'ci/common-ci.yml'
+
+rocm clang14+rocm-5.3.3 release deps:
```
Comment (msimberg) [inline ci/rocm/clang14+rocm-5.3.3_release.yml:4]: See other comment about naming the file. Same comment applies to the job name.

**Review by msimberg — APPROVED** [review]
This looks nice and clean to me! A couple of superficial comments, nothing blocking.

File: ci/common-ci.yml (hunk 8 above)
Comment (rasolca) [reply ci/common-ci.yml:86]: Your comment triggered a couple of questions:

glibc-tools:
As we now we use ubuntu 22.04 based images for all tests can we move it to the dockerfile even if ubuntu20.04 is still supported?

jq strace: 
Why are they not installed for all the pipelines and therefore added to the dockerfile?

Comment (msimberg) [reply ci/common-ci.yml:86]: Good questions. I see no reason not to simply always install them in the deploy images. I think I may originally have thought that `jq` and `strace` wouldn't be needed in all the pipelines, but really there's not much harm in keeping them there always.

So :heavy_check_mark: from to move them all to the dockerfile.

**Comment by rasolca** [comment]
cscs-ci run

**Comment by rasolca** [comment]
cscs-ci run

**Comment by rasolca** [comment]
cscs-ci run

**Comment by rasolca** [comment]
cscs-ci run

**Comment by rasolca** [comment]
cscs-ci run

**Comment by rasolca** [comment]
cscs-ci run

**Pull request merged**